array expression work

schweitzpgi opened the pull request.

Commit:
Array expression work.

Commit:
changes to implement the basic transformation. this does not remove the
array_load yet.

Commit:
changes

jeanPerier committed:
Get length of assumed length named constant in mlir type

jeanPerier committed:
Remove old TypeBuilder and simplify type translation

- Change all entry points to use the type builder that does not
  visit the expression tree but uses GetSahpe/LEN() + Fold instead.
- Remove unused type translation entry points.
- Remove some template arguments that can now be handled by using
  different function name (they used explicit template instantiation
  in the remaining usages, which is less readable).
- Set the location in error messages when lowering symbol types.
- Lower typeless expressions to fir type (function pointer
  to their fir types, the rest to fir.ref<NoneType>/NoneType).

Commit:
changes: get array.f90 working again and connect some of the array ops.

schweitzpgi committed:
Fix the CFG rewriting pass to properly handle the iterate_while when it
has negative step (count down) as well as positive.

Commit:
changes

jeanPerier reviewed (commented):
I really like the design of the added fir ops, they are rather easy to understand.
It looks like the patch you have will work great with "full" elemental expression (where every node of the expression is an elemental operator/function and where array symbols are referenced without indexes/slices).

I think  the general case of array expressions will require some more complexity with the array context and some analysis of the expressions nodes before assuming they are elemental. (I think you know that already and the purpose of this PR is not to cover all array expr :) )
The kind of next level design Fortran I can think of are:
```
real, dimension(100) :: A, B, C
real, external :: foo ! scalar function taking an array argument
real, external :: bar ! simple scalar function
A = B + foo(sin(A) + C(n) + bar()) + C
```

I'll try to think how best I would use what you have here to do that while going to the kindergarten :)

schweitzpgi committed:
Fix the bounds check on iterate_while. The upper bound is defined to be
inclusive. (The iteration set is a closed iterval [lo, lo+step, ... hi],
where hi is included if it is equal to lo+k*step for some integer k)

schweitzpgi commented on flang/lib/Lower/ConvertExpr.cpp:
```diff
@@ -291,19 +313,22 @@ class ExprLowering {
+            loc, eleTy, base, exprCtx.getLoopCounters());
+        return fetch.getResult();
+      }
+      return genLoad(fir::getBase(exv));
```
Yep. It's a known issue at the moment to lower the scalar subexpressions correctly. (They are lowered, but not used in the loop nest yet.)

schweitzpgi commented on flang/lib/Lower/ConvertExpr.cpp:
```diff
@@ -291,19 +313,22 @@ class ExprLowering {
+      // `sym` appears in an array expression. This is the code gen phase inside
+      // the loop nest.
+      auto exv = toExtendedValue(symMap.lookupSymbol(sym));
+      if (fir::getBase(exv).getType().isa<fir::SequenceType>()) {
```
I, of course, like the type-based solution. 👍 I wish I had done it from the beginning...

I agree that with all the cleanup that you've done, and since it is fresh in your memory, now might be the best time to transition and replace the `!fir.char<KIND>` type with a string type `!fir.char<LEN, KIND>`.

schweitzpgi commented on flang/lib/Lower/Bridge.cpp:
```diff
@@ -1398,89 +1398,137 @@ class FirConverter : public Fortran::lower::AbstractConverter {
+  /// 1. Evaluate the lhs to determine the rank and how to form the ArrayLoad
+  /// (e.g., if there is a slicing op).
+  /// 2. Scan the rhs, creating the ArrayLoads and evaluate the scalar subparts
+  /// to be added to the map.
```
You're right. This is a hole.

In fact, I want to track these and stick them someplace like the SymbolMap ... but they are not symbols, they are just lowered scalar expressions. I think I will extend the SymbolMap to allow "temporary" bindings. That should be a good general solution applicable to other constructs, etc.

jeanPerier replied:
The addresses of the `evaluate::Expr` under the parse-tree should be stable, so it may be possible to use these to create a binding here, but that may be a bit risky to assume we do not generate and lower a temporary evaluate::Expr at some point...

A solution could also be to keep a single pass and generate code in both the prelude/loop according to the needs.

We could maybe stick under the Expression entry node (that should also be used while entering operands sub-expressions during the visit) something like:
```
fir::ExtendedValue genval(SomeExpr expr) {
  fir::ExtendedValue res;
  if (inArrayContext()) {
   if (isArray(expr))
     context.generateCodeInLoopNest(); // set builder insertion point
     return genVisit(expr); // lgenVisit would just be like the current genval(SomeExpr)
   else // scalar
     context.generateCodeInPrelude();
     // we are not in an array context for this operand, but we do not want to pop the
    // array context that we still need for the other operands.
     context.setNotInArrayContext(); 
     res = genVisit(expr);
  } else {
    // we are not in array context but if the Expression we want to lower is an elemental operation/function and the expr
    // has an array shape, we need to create a new array context here. This cover for instance the case where a scalar 
    // function is taking an array argument (foo(A + B)) and we need an array context for A + B. 
    if (isElementalExpr(expr)) {
      // we probably want to create a temp array here that can be assigned to, and 
      auto tempArray = context.generateArrayTempStorage(expr);
      context.AddTempToCleanUpList(tempArray);
      context.createNewArrayContext(expr, tempArray);
      auto res = genVisit(expr);
      // pop the array context, the temp cannot be deleted until it has been used;
      context.popArrayContext();
    }
  }
 // clean up the temps created by lower level operand if any (this would be the tempArray that may have been
 //    generated by nested call to `gen(SomeExpr)`, but not the one that we may have just generated in the current
 // call (it's not used yet)).
 context.CleanUpLowerLevelTemp();
}
```

Doing so may also help with the fact that expressions may have several elemental array sub-expressions that should not live in the same array context.

Then, in gen(symbolRef) we would also need to be modified to generate the array loads in the prelude insertion point, and the ref in the loop nest at the same time.

This is just an idea based on the observation that these visits are a central nerve of expression lowering where we could maybe centralized the array expression aspect: https://github.com/flang-compiler/f18-llvm-project/blob/c10d6c987e7d6d341a5486589e2d84483fdb3688/flang/lib/Lower/ConvertExpr.cpp#L1611-L1618
But it may make more sense to deploy some sort of common handling in the expressions node that are elemental instead.

schweitzpgi replied:
I'm reluctant to push too much of the smarts into the expression helper. I don't think that's the right place for it after all. I'm going to experiment a bit...

schweitzpgi commented on flang/lib/Lower/SymbolMap.h:
```diff
@@ -253,10 +263,11 @@ class SymMap {
   }
 
-  llvm::DenseMap<const semantics::Symbol *, SymbolBox> symbolMap;
+  llvm::SmallVector<llvm::DenseMap<const semantics::Symbol *, SymbolBox>, 4>
```
Agreed

schweitzpgi commented on flang/lib/Lower/SymbolMap.cpp:
```diff
@@ -0,0 +1,170 @@
+      return iter->second;
+  }
+  // FIXME BUG: How does this know that the caller is expecting the host
+  // associated symbol?
```
I think I want the distinction to be forced on the client though.

`lookupSymbol(x)` logically should lookup the symbol `x` and nothing more.

On the other hand, a `lookupAssociatedSymbol(x)` would lookup either `x` or whatever `x` was associated with. I don't know if there will ever be a case where the former is ever required, but I'm not sure I like conflating these two things by default.

jeanPerier replied:
I understand your point. All the places I can think of would need to use `lookupAssociatedSymbol` as of today.
Two alternatives are:
a) add a boolean flag to disable the fallback to host associated symbols for the clients where it would not make sense to follow host association (since in my opinion, these should be the weirdo cases).
b) create SymbolBox for HostAssociated symbols in any cases and remove this fallback. That may just be the cleanest solution. Only drawback is that it may lead to a bigger symbol map, but I do not think there should be too many of those symbols where we would create a SymbolBox for nothing (only Blocks/Stmt functions and maybe OpenMp).

schweitzpgi replied:
If the 99% case is that a symbol marked as "host associated" really means the variable it is "host assoicated" with then that semantics should be made explicit in the binding (i.e., (b)). Probing the table with a magic error recovery side effect is muddy. With the current approach, it's unclear if the symbol *must* be in the table, if the lookup is for a particular symbol *only* and where the context should be aware if the symbol shadows another variable, if a symbol isn't found whether the error is that missing symbol or some other missing symbol, etc.

schweitzpgi commented on flang/lib/Lower/ConvertExpr.cpp:
```diff
@@ -315,19 +340,29 @@ class ExprLowering {
+      // `sym` appears in an array expression. This is the prelude phase where
+      // we capture immutable values to be used inside the loop nest.
+      auto memref = fir::getBase(var);
+      if (fir::isArray(var)) {
```
Yep. I'm going to have to tune this some to work with slices, do proper scalar promo, etc.

schweitzpgi commented on flang/lib/Lower/ConvertExpr.cpp:
```diff
@@ -291,19 +313,22 @@ class ExprLowering {
+      if (fir::getBase(exv).getType().isa<fir::SequenceType>()) {
+        auto base = fir::getBase(exv);
+        auto eleTy = base.getType().cast<fir::SequenceType>().getEleTy();
+        auto fetch = builder.create<fir::ArrayFetchOp>(
```
You're right. I'm debating here about keeping an explicit context stack or to maintain the stack on traversal. I'm leaning towards just making it explicit in the helper.

schweitzpgi committed:
Merge pull request #544 from schweitzpgi/ch-countdown-fix

Fix the CFG rewriting pass to properly handle the iterate_while when it

schweitzpgi commented:
@jeanPerier I don't see any problems in fcvs from this. Do you think this is good enough to move into fir-dev as is? There are some obvious things that still need doing, but this is getting rather huge already.

jeanPerier commented:
> I don't see any problems in fcvs from this. Do you think this is good enough to move into fir-dev as is? There are some obvious things that still need doing, but this is getting rather huge already.

It's fine by me. The fir/optimizer part looks clean enough in my opinion. It's clear that the bridge part will create wrong compile/crashes for some F95 programs, but this code should only kick in for array assignments and I think we were already failing those anyway from what I can test in my sandbox.

Commit:
Add bbc entry in lit tests

Fix indentation

jeanPerier committed:
Merge pull request #541 from flang-compiler/jpr-type-simplification

Simplify type lowering and reflect constant length in more cases

schweitzpgi committed:
Merge pull request #547 from Rin18/Rin/add_bbc_entry

Add bbc entry in lit tests

schweitzpgi committed:
changes: remove erase

schweitzpgi commented:
Closing because of the rebasing fun. Thanks for your helpful thoughts.
